packages/grid/x-data-grid-pro/src/hooks/features/lazyLoader/useGridLazyLoader.ts

```diff
@@ -0,0 +1,179 @@
+import { getRenderableIndexes } from '@mui/x-data-grid/hooks/features/virtualization/useGridVirtualScroller';
+import { GridApiPro } from '../../../models/gridApiPro';
+import { DataGridProProcessedProps } from '../../../models/dataGridProProps';
+import { GRID_SKELETON_ROW_ROOT_ID } from './useGridLazyLoaderPreProcessors';
+import { GridFetchRowsParams } from '../../../models/gridFetchRowsParams';
+
+function findSkeletonRowsSection(
+  visibleRows: GridRowEntry<any>[],
+  range: { firstRowIndex: number; lastRowIndex: number },
+): { firstRowIndex: number; lastRowIndex: number } {
+  const visibleRowsSection = visibleRows.slice(range.firstRowIndex, range.lastRowIndex);
+  let firstRowIndex = range.firstRowIndex;
+  let lastRowIndex = range.lastRowIndex;
+  let startIndex = 0;
+  let endIndex = visibleRowsSection.length - 1;
+  let isSkeletonSectionFound = false;
+
+  while (!isSkeletonSectionFound) {
+    if (!visibleRowsSection[startIndex].model && !visibleRowsSection[endIndex].model) {
+      isSkeletonSectionFound = true;
+    }
+
+    if (visibleRowsSection[startIndex].model) {
+      startIndex += 1;
+      firstRowIndex += 1;
+    }
+
+    if (visibleRowsSection[endIndex].model) {
+      endIndex -= 1;
+      lastRowIndex -= 1;
+    }
+  }
+
+  return {
+    firstRowIndex,
+    lastRowIndex,
+  };
+}
+
+/**
+ * @requires useGridRows (state)
+ * @requires useGridPagination (state)
+ * @requires useGridDimensions (method) - can be after
+ * @requires useGridScroll (method
+ */
+export const useGridLazyLoader = (
+  apiRef: React.MutableRefObject<GridApiPro>,
+  props: Pick<
+    DataGridProProcessedProps,
+    'onFetchRows' | 'rowsLoadingMode' | 'pagination' | 'paginationMode' | 'rowBuffer'
+  >,
+): void => {
+  const visibleRows = useGridVisibleRows(apiRef, props);
+  const rowIds = useGridSelector(apiRef, gridVisibleSortedRowIdsSelector);
+  const sortModel = useGridSelector(apiRef, gridSortModelSelector);
+  const filterModel = useGridSelector(apiRef, gridFilterModelSelector);
+  const renderedRowsIntervalCache = React.useRef<GridRenderedRowsIntervalChangeParams>({
+    firstRowToRender: 0,
+    lastRowToRender: 0,
+  });
+
+  const handleRenderedRowsIntervalChange = React.useCallback(
+    (params: GridRenderedRowsIntervalChangeParams) => {
+      const dimensions = apiRef.current.getRootDimensions();
+
+      if (
+        !dimensions ||
+        props.rowsLoadingMode !== GridFeatureModeConstant.server ||
+        (renderedRowsIntervalCache.current.firstRowToRender === params.firstRowToRender &&
+          renderedRowsIntervalCache.current.lastRowToRender === params.lastRowToRender)
+      ) {
+        return;
+      }
+
+      const renderedRowsIds: Array<GridRowId | null> = [...rowIds].splice(
+        params.firstRowToRender,
+        params.lastRowToRender - params.firstRowToRender,
+      );
+      const hasSkeletonRowIds = renderedRowsIds.some(
+        (rowId) => `${rowId}`.indexOf(GRID_SKELETON_ROW_ROOT_ID) >= 0,
```

Comment:
By the way, instead of `some`, you can use `findIndex` such that you get at the same time the start index of `skeletonRowSection`. And to get the last one you can use `findLastIndex`